### PR TITLE
Prevent flattening multiplicative operations

### PR DESCRIPTION
**Pull request by duailibe**
Context: https://github.com/prettier/prettier/issues/187#issuecomment-385495801

### CODE REVIEW

**Comment by lydell** [comment]
So, no previous snapshots changed? That’s sad. Would have helped to see how this changes things. Either way we need to run this on a couple of projects.

File: tests/binary_math/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -22,4 +26,10 @@ sign = 1 - 2 * (b[3] >> 7);
 exponent = (((b[3] << 1) & 0xff) | (b[2] >> 7)) - 127;
 mantissa = ((b[2] & 0x7f) << 16) | (b[1] << 8) | b[0];
 
+((2 / 3) * 10) / 2 + 2;
```
[the remark this answers is not in the document]
Comment (duailibe) [reply tests/binary_math/__snapshots__/jsfmt.spec.js.snap]: Good news is if your input is `(2 / 3) * (10 / 2) + 2` it outputs `(2 / 3) * (10 / 2) + 2`

[the remark this answers is not in the document]
Comment (lydell) [reply tests/binary_math/__snapshots__/jsfmt.spec.js.snap]: Yeah, we preserve the AST.

**Comment by duailibe** [comment]
@lydell a few have changed, I forgot to commit them 😓

**Comment by lydell** [comment]
Here are most of the changes on ~60k code:

17k+42k+20k=60k

```diff
-    const cartSummaryY = -this.NAV_HEADER_HEIGHT / 2 * this.progress;
-    const cartSummaryScale = 1 - 1 / 3 * this.progress;
+    const cartSummaryY = (-this.NAV_HEADER_HEIGHT / 2) * this.progress;
+    const cartSummaryScale = 1 - (1 / 3) * this.progress;

-    const cartTextWrapperScale = Math.max(1, 1 + 1 / 3 * this.progress);
+    const cartTextWrapperScale = Math.max(1, 1 + (1 / 3) * this.progress);

   const boxWidth =
-    DEFAULT_WIDTH_MM / DEFAULT_MM_PER_WIDTH_PERCENTAGE * rect.width;
+    (DEFAULT_WIDTH_MM / DEFAULT_MM_PER_WIDTH_PERCENTAGE) * rect.width;

     Math.min(
       DEFAULT_MAX_WIDTH_PERCENTAGE * ratio,
-      region.height / DEFAULT_MM_PER_WIDTH_PERCENTAGE * ratio
+      (region.height / DEFAULT_MM_PER_WIDTH_PERCENTAGE) * ratio
     )

   const updatePoint = (point, xDelta, yDelta) => ({
-    x: clamp(centerX + width / 2 * xDelta),
-    y: clamp(centerY + height / 2 * yDelta),
+    x: clamp(centerX + (width / 2) * xDelta),
+    y: clamp(centerY + (height / 2) * yDelta),
   });

-  const leftRightMax = Math.PI * 3 / 2;
+  const leftRightMax = (Math.PI * 3) / 2;

-  const ipDistance = ibDistance * MIN_CORNER_DISTANCE_EDGE_MOVE / bcDistance;
+  const ipDistance = (ibDistance * MIN_CORNER_DISTANCE_EDGE_MOVE) / bcDistance;

-  const k = dy * height / (dx * width);
+  const k = (dy * height) / (dx * width);

-      lng: position.lng() - mapWidth * this.searchAreaRatio() / 2,
+      lng: position.lng() - (mapWidth * this.searchAreaRatio()) / 2,

-            cx={LINE_MARGIN + n * (width - LINE_MARGIN * 2) / numSegments}
+            cx={LINE_MARGIN + (n * (width - LINE_MARGIN * 2)) / numSegments}

           {
             x:
               LINE_MARGIN +
-              differenceInCalendarMonths(round.date, firstDate) *
-                (width - LINE_MARGIN * 2) /
+              (differenceInCalendarMonths(round.date, firstDate) *
+                (width - LINE_MARGIN * 2)) /
                 numSegments,
             y: lineY,

-                style={{ left: `${(value - min) / (max - min) * 100}%` }}
+                style={{ left: `${((value - min) / (max - min)) * 100}%` }}

-        ? formatNumber(100 * total / assessed)
+        ? formatNumber((100 * total) / assessed)

-    return a * b * c / (v * (a * b + a * c + b * c));
+    return (a * b * c) / (v * (a * b + a * c + b * c));
```

**Comment by lydell** [comment]
@duailibe @vjeux What do you think about turning `a % 10 - 5` into `(a % 10) - 5`? Different PR?

**Comment by duailibe** [comment]
@lydell Yeah we can do in another PR (it will change in a different part of the code as well)

**Pull request merged**